### DISCUSSION TITLE
Proper way to control the actuator via python?

### DISCUSSION THREAD

**Discussion by DrawZeroPoint** [google-deepmind/mujoco#237] [Asking for Help] [answered]
Dear MuJoCo developers,

I am using the 2.1.4 version python binding of MuJoCo to control actuators with codes. So far, I encountered an odd issue that may be due to my improper way to use the API, so any suggestions are gratefully appreciated.

The simplified code I am using is like this:

```
actuator = self._data.actuator(actuator_name)
actuator.ctrl = cmd
```

where I can confirm that `_data` is initialized and contains the actuator;  `cmd` is a plain float number. Oddly, the second line gives the error:

`IndexError: too many indices for array: array is 0-dimensional, but 1 were indexed`

I wonder if it is due to `.ctrl` should not be used in this way? What I planned to do is to assign the control value `cmd` to the actuator.
Previously with mujoco-py, I did the same thing like this:

```
self.sim.data.ctrl[actuator_id] = cmd
```

**Comment by JeanElsner** [comment]
I haven't really used the named accessors yet, but the ctrl field of the actuator view may not have been intended to be written into.
What I usually do is that I keep a list or map of indices that I initialise once like this:
```
id = mujoco.mj_name2id(model, mujoco.mjtObj.mjOBJ_ACTUATOR, name)
```
You can then assign values to the ctrl field like you used to, i.e.
```
data.ctrl[id] = ctrl_d
```

**Comment by saran-t** [comment, marked as answer]
You are using it correctly (or at least how I intended for it to be used). This is a bug in the Python bindings. I'm going to create an issue to track this bug, thanks for bringing it up!

**Reply by saran-t** [reply]
This should now be fixed in 55c1c9138293e1c96770916f47d2479b4cb2a662. You'll have to build it manually from source for now, though, until I get around to releasing a new version.

**Reply by saran-t** [reply]
The fixed has been incorporated into version 2.1.5 that was released today.

**Comment by DrawZeroPoint** [comment]
Thank you @saran-t for the quick fix! As your accepted answer, I followed [Building from source](https://mujoco.readthedocs.io/en/latest/python.html#building-from-source), and here are some sidenotes I think might be helpful for revealing some details not addressed in the doc:

1. Inside the cloned ~/mujoco folder (a Python virtual environment needs to be **activated** to run make_sdist.sh):

  ```
  cd python && ./make_sdist.sh
  ```

2. Then, set the env variable pointing to the **binary** of MuJoCo (here showcases my setup):

  ```
  export MUJOCO_PATH=/home/dzp/.mujoco/mujoco214
  ```

3. Run `pip wheel name_of_sdist.tar.gz`, where `name_of_sdist.tar.gz` is generated in the last step inside `mujoco/pyhon/dist`

  ```
  cd dist && pip wheel mujoco-2.1.4.tar.gz
  ```

4. Finally, install the wheel (you may need to prune the previous installation done by `pip install mujoco`):

  ```
  pip install mujoco-2.1.4-cp38-cp38-linux_x86_64.whl
  ```

My work environment is Ubuntu 20.04 with Python 3.8. The aforementioned steps lead me successfully solved the issue.

Besides, I confirmed that @JeanElsner's circumvention is helpful for those who prefer an official release, thank you as well.
